File: python/linkedlist/leetcode/linkedlist_cycle.py

# 判断链表是否有环
# https://leetcode-cn.com/problems/linked-list-cycle

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:

    # 使用一个set
    def hasCycle_1(self, head: ListNode) -> bool:
        s = set()
        while head:
            if head in s:
                return True
            s.add(head)
            head = head.next
        return False

    # 快慢指针
    def hasCycle_2(self, head: ListNode) -> bool:
        slow, fast = head, head
        while fast and fast.next:
            slow, fast = slow.next, fast.next.next
            if slow == fast:
                return True
        return False

File: go/linkedlist/leetcode/linked_list_cycle.go
package main

//环形链表

//使用set判重
func hasCycle1(head *ListNode) bool {
	set := map[*ListNode]bool{}
	for head != nil {
		if _, ok := set[head]; ok {
			return true
		}
		set[head] = true
		head = head.Next
	}
	return false
}

//快慢指针法
func hasCycle2(head *ListNode) bool {
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow, fast = slow.Next, fast.Next.Next
		if slow == fast {
			return true
		}
	}
	return false
}
